R/verify.R

```diff
@@ -0,0 +1,53 @@
+#' Check Code Reproducibility
+#'
+#' Checks whether code in `teal_data` reproduces the stored objects.
+#'
+#' @return If the objects returned by the `@code` slot in the input `teal_data` object exactly match
+#' the objects in `teal_data@env`, the function updates the `@verified` slot to `TRUE` in the `teal_data` object.
+#' If the `@code` fails to recreate objects in `teal_data@env`, an error is thrown.
+#'
+#' @param x `teal_data` object
+#' @examples
+#' tdata1 <- teal_data()
+#' tdata1 <- within(tdata1, {
+#'   a <- 1
+#'   b <- a^5
+#'   c <- list(x = 2)
+#' })
+#' verify(tdata1)
+#'
+#' tdata2 <- teal_data(x1 = iris, code = "x1 = iris")
+#' verify(tdata2)
+#' verify(tdata2)@verified
+#' tdata2@verified
+#'
+#' \dontrun{
+#' tdata3 <- teal_data()
+#' tdata3 <- within(tdata3, {
+#'   stop("error")
+#' })
+#' }
+#' @name verify
+#' @aliases verify,teal_data-method
+#' @aliases verify,qenv.error-method
+#' @rdname verify
+#' @export
+setGeneric("verify", function(x) standardGeneric("verify"))
+setMethod("verify", "teal_data", definition = function(x) {
+  if (x@verified) {
+    return(x)
+  }
```

Comment:
Here we return `x` right away if `x@verified` is TRUE.

However, user can manually assign value to this slot.

Here's an example, let's start with creating `teal_data` object where I purposely provide the wrong code for my data:

<img src="https://github.com/insightsengineering/teal.data/assets/8597300/72e5fe9a-e661-4429-9ef7-a46fad408646" width="50%" height="50%" />

If I set the `td@verified` slot manually to `TRUE`, I am expecting the `verify()` will overwrite the slot to `FALSE`, but instead I got `TRUE` because of the current return condition.

<img src="https://github.com/insightsengineering/teal.data/assets/8597300/1fc47908-760e-4476-8bc0-fe64c9d7ecca" width="50%" height="50%" />

I guess it's highly unlikely someone will manually overwrite the `verified` slot but since it's possible to do so, I don't think it's a bad idea to set/reset the value when `verify()` is executed.

Or, is there a way to protect a slot to prevent user assignment?

Reply:
This is in line with the intent as stated in the function's documentation.
1. create data
2. verify data
3. use data in app

If your code is not reproducible, you know before you run the app. If you still run the app with data like that, you will get a warning added to your code.

It is impossible to prevent manual setting of slots. If one wants to cheat, one will find a way.

Reply:
Thanks for the explanation.

I do agree with the view, and furthermore, we do have this language in the teal_data-class documentation stating:
> These slots should never be accessed directly, use the provided get/set functions.

Since we don't have a setter for `verified`, it should be somewhat apparent that this shouldn't be touch by user.
I also like that we don't expose `verified` slot in the `teal_data` constructor.

@chlebowa 
About step 2 that you mentioned `verify data`, is this something that we will handle in the back-end in `init()`or do we ask user to run `verify()` before running `init()`?

Reply:
The latter.